### PR TITLE
Fixes CircularAppender out of order message bug

### LINKED ISSUE #983: Usage of circular buffer in network test is wrong
I'm seeing the following in the log files:

```
2020-06-25 16:33:42,114 Trace [agora.network.NetworkClient] - Request 'getBlockHeight' to Validator #1 (GDB2...DGJJ) failed: Connection with peer closed
2020-06-25 16:33:42,214 Trace [agora.network.NetworkClient] - Request 'getBlockHeight' to Validator #1 (GDB2...DGJJ) failed: Connection with peer closed
2020-06-25 16:33:42,319 Trace [agora.network.NetworkClient] - Request 'getBlockHeight' to Validator #1 (GDB2...DGJJ) failed: Connection with peer closed
2020-06-25 16:33:42,422 Trace [agora.network.NetworkClient] - Request 'getBlockHeight' to Validator #1 (GDB2...DGJJ) failed: Connection with peer closed
2020-06-25 16:33:42,528 Trace [agora.network.NetworkClient] - Request 'getBlockHeight' to Validator #1 (GDB2...DGJJ) failed: Connection with peer closed
2020-06-25 16:33:43,583 Info [agora.network.NetworkManager] - Doing network discovery..:33:28,21 Trace [agora.node.FullNode] - Received Transaction: { value: { type: const(TxType).Payment, inputs: [{ previous: 0x6a69b9d6485149b3dcf50e5577bdeb2021580507d2e9237eb7cf702285004fb93eaddca22f7dcfca11f1a20012f7ca72e26fc02c3a5d4b7afac5625cc5a362c8, index: 0, signature: 0x068a6b96e0b42d2ccfb95c13fe11cc1b390edcde57a3e4134c6594bcca18703abf80b2008378f21dad904d98f04822b43f20f1a9cc619908e2cdf7ef034a9a61 }], outputs: [{ value: 625000000000000, address: GDQ22X67LUSWO4TN6D5KHXBQDBW43PDHI527ITKYK3Q6T3H5ZGXCAVEO }] } }
2020-06-25 16:33:28,21 Trace [agora.node.FullNode] - Received Transaction: { value: { type: const(TxType).Payment, inputs: [{ previous: 0x9bef6129d38e26579ed4dd9a3f49143c40dcb85f7e93365a4cf6b1196023107f49b6444e3f780fc55f7a64809cdcde77aa520331242bb44dc063be7cc2ec614a, index: 0, signature: 0x02bb813645f604746eedf2075387ee0dc9d45c57d1fcf5e4cb5e20c83e2bb944c3a5c66e2ea2d2151b1093fac1aec9f81f6465b768c1e9620d963907be3c85b8 }], outputs: [{ value: 625000000000000, address: GDQ22X67LUSWO4TN6D5KHXBQDBW43PDHI527ITKYK3Q6T3H5ZGXCAVEO }] } }
2020-06-25 16:33:28,22 Trace [agora.node.FullNode] - Received Transaction: { value: { type: const(TxType).Payment, inputs: [{ previous: 0x11d681527ccac0c6e8b2e25a6ea642fdb3222dca72396f1db2ccc268f8b36a45429e5570b6d6d40aeadcfb28c210ab368d2bcad52ca32c7660ddd2f4d95d5ed2, index: 0, signature: 0x09686e5033984b9e9e3ac4a059c74c807ef8b5df9bba76f1d8cf180986eb13e3bce0a45cf74831e2d5ae226c57ad3d8a9a5f2b1febe98ad4d4da5687e3baf11e }], outputs: [{ value: 625000000000000, address: GDQ22X67LUSWO4TN6D5KHXBQDBW43PDHI527ITKYK3Q6T3H5ZGXCAVEO }] } }
```

Obviously that's wrong, we shouldn't print the buffer as-is. I probably introduced this bug, too.

### PR DESCRIPTION
**Pull request by ferencdg**
fixes #983

### CODE REVIEW

File: source/agora/utils/Log.d [hunk 1]
```diff
@@ -156,7 +164,7 @@ public class CircularAppender : Appender
     {
         // edge-case: if the buffer isn't filled yet,
         // write from buffer index 0, not the cycle's current index
-        if (this.used_length <= this.buffer.length)
```
Comment (ferencdg) [inline source/agora/utils/Log.d:159]: This was the actual bug, this.used_length is ALWAYS <= this.buffer.length, so the else branch never executed.

The this.used_length == this.buffer.length case is handled exactly same for the if and the else branch.

File: source/agora/utils/Log.d [hunk 2]
```diff
@@ -198,6 +206,49 @@ public class CircularAppender : Appender
     }
 }
 
+///
+unittest
+{
+    class MockLayout : Appender.Layout
```
Comment (Geod24) [inline source/agora/utils/Log.d]:
```suggestion
    static class MockLayout : Appender.Layout
```

File: source/agora/utils/Log.d [hunk 3]
```diff
@@ -198,6 +206,49 @@ public class CircularAppender : Appender
...
+    {
+        /// Format the message
+        public override void format (LogEvent event, scope FormatterSink dg)
+        {
+            sformat(dg, "{}", event);
+        }
+    }
+
+    auto get_log_output = (const string log_msg){
```
Comment (Geod24) [inline source/agora/utils/Log.d]:
```suggestion
    scope get_log_output = (string log_msg){
```

`const string` implicitly converts to `string`, and the other way around, so no need to clog the prototype with a constraints that will only apply to the function's body.

File: source/agora/utils/Log.d [hunk 4]
```diff
@@ -111,16 +111,25 @@ public alias Log = Ocean.Log;
...
 
+public CircularAppender!BufferSize getCircularAppender (size_t BufferSize = 2^^20)()
+{
+    static CircularAppender!(BufferSize) appender;
+        if (appender is null)
+            appender = new CircularAppender!(BufferSize)();
+
+    return appender;
+}
+
```
Comment (Geod24) [inline source/agora/utils/Log.d]: Why the `opCall` removal ? I understand you had to template the class for testing purpose, and it makes sense, but you could just put the default value on the `class` and change the calls from `CircularBuffer()` to `CircularBuffer!()()`.

File: source/agora/utils/Log.d [hunk 5]
```diff
@@ -111,16 +111,25 @@ public alias Log = Ocean.Log;
...
 /// Circular appender which appends to an internal buffer
-public class CircularAppender : Appender
+public class CircularAppender(size_t BufferSize) : Appender
```
Comment (Geod24) [inline source/agora/utils/Log.d]:
```suggestion
public class CircularAppender (size_t BufferSize) : Appender
```

Space before parenthesis on definition

File: source/agora/utils/Log.d [hunk 6]
```diff
@@ -198,6 +206,49 @@ public class CircularAppender : Appender
...
+
+        immutable buffer_size = 7;
+        char[buffer_size + 1] result_buff;
+        LogEvent event;
+        auto appender = getCircularAppender!(buffer_size)();
+        appender.clear();
+        event.set(ILogger.Context.init, ILogger.Level.init, log_msg, "");
+        appender.layout(new MockLayout());
+        appender.append(event);
+        appender.print(result_buff[]);0 .. min(log_msg.length, buffer_size)]);
```
Comment (Geod24) [inline source/agora/utils/Log.d]: I don't think the compiler will agree with you on this one

File: source/agora/utils/Log.d [hunk 7]
```diff
@@ -198,6 +206,49 @@ public class CircularAppender : Appender
...
+    }
+
+    auto get_log_output = (const string log_msg){
+        import ocean.util.log.ILogger;
+
+        immutable buffer_size = 7;
+        char[buffer_size + 1] result_buff;
+        LogEvent event;
+        auto appender = getCircularAppender!(buffer_size)();
+        appender.clear();
```
Comment (Geod24) [inline source/agora/utils/Log.d]:
```suggestion
        scope appender = new CircularAppender!(buffer_size);
```

File: source/agora/utils/Log.d [hunk 8]
```diff
@@ -198,6 +206,49 @@ public class CircularAppender : Appender
...
+
+    // case 1
+    // the internal buffer size is greater than the length of the messages 
+    // that we are trying to log
+    assert(get_log_output("01234") == "01234");
+
+    // case 2
+    // the internal buffer size is equal to the length of the messages 
+    // that we are trying to log(there is a terminating newline)
+    assert(get_log_output("012345") == "012345");
```
Comment (Geod24) [inline source/agora/utils/Log.d:235]: Shouldn't this be:
```suggestion
    assert(get_log_output("012345") == "012345\n");
```
?

**Review by Geod24 — CHANGES_REQUESTED** [review]
Few nits, typical OB1, nice catch!

File: source/agora/utils/Log.d (hunk 8 above)
Comment (ferencdg) [reply source/agora/utils/Log.d:235]: get_log_output doesn't return the newline, let me know if you want me to change this

File: source/agora/utils/Log.d (hunk 4 above)
Comment (ferencdg) [reply source/agora/utils/Log.d]: I think I just didn't find CircularBuffer!()() pretty, but I changed it now

Comment (Geod24) [reply source/agora/utils/Log.d]: Yeh it's not. Renaming `opCall` to `get` would also work, but it should be a separate commit at least.

File: source/agora/utils/Log.d (hunk 8 above)
Comment (Geod24) [reply source/agora/utils/Log.d:235]: Missed that. Works for me.

**Review by Geod24 — APPROVED** [review]

File: source/agora/utils/Log.d [hunk 9]
```diff
@@ -143,9 +143,9 @@ public class CircularAppender : Appender
 
     public static CircularAppender opCall ()
     {
-        static CircularAppender appender;
+        static CircularAppender!BufferSize appender;
```
Comment (Geod24) [inline source/agora/utils/Log.d:146]: Protip: Inside a template, you can use the template name to refer to the instantiation. That's why you didn't have to change this function's return value.

Comment (ferencdg) [reply source/agora/utils/Log.d:146]: got it thanks

**Pull request merged**